transports/wasm-ext/src/webtransport/substream.rs

```diff
@@ -0,0 +1,190 @@
+use futures::{AsyncRead, AsyncWrite};
+use js_sys::Uint8Array;
+use send_wrapper::SendWrapper;
+use std::io;
+use std::pin::Pin;
+use std::task::ready;
+use std::task::{Context, Poll};
+use web_sys::{
+    ReadableStreamDefaultReader, WebTransportBidirectionalStream, WritableStreamDefaultWriter,
+};
+
+use crate::webtransport::cached_js_promise::CachedJsPromise;
+use crate::webtransport::utils::{detach_promise, parse_reader_response, to_io_error, to_js_type};
+use crate::webtransport::Error;
+
+pub struct Substream {
+    reader: SendWrapper<ReadableStreamDefaultReader>,
+    reader_read_promise: CachedJsPromise,
+    read_leftovers: Option<SendWrapper<Uint8Array>>,
+    writer: SendWrapper<WritableStreamDefaultWriter>,
+    writer_state: StreamState,
+    writer_ready_promise: CachedJsPromise,
+    writer_closed_promise: CachedJsPromise,
+}
+
+#[derive(PartialEq, Eq)]
+enum StreamState {
+    Open,
+    Closing,
+    Closed,
+}
+
+impl Substream {
+    pub(crate) fn new(bidi_stream: WebTransportBidirectionalStream) -> Result<Self, Error> {
+        let recv_stream = bidi_stream.readable();
+        let send_stream = bidi_stream.writable();
+
+        let reader = to_js_type::<ReadableStreamDefaultReader>(recv_stream.get_reader())?;
+        let writer = send_stream.get_writer()?;
+
+        Ok(Substream {
+            reader: SendWrapper::new(reader),
+            reader_read_promise: CachedJsPromise::new(),
+            read_leftovers: None,
+            writer: SendWrapper::new(writer),
+            writer_state: StreamState::Open,
+            writer_ready_promise: CachedJsPromise::new(),
+            writer_closed_promise: CachedJsPromise::new(),
+        })
+    }
+
+    fn poll_writer_ready(&mut self, cx: &mut Context) -> Poll<io::Result<()>> {
+        if self.writer_state != StreamState::Open {
+            return Poll::Ready(Err(io::ErrorKind::BrokenPipe.into()));
+        }
+
+        let desired_size = self
+            .writer
+            .desired_size()
+            .map_err(to_io_error)?
+            .map(|n| n.trunc() as i64)
+            .unwrap_or(0);
+
+        // We need to poll if the queue is full or if the promise was already activated
+        if desired_size <= 0 || self.writer_ready_promise.is_active() {
```

Comment:
`desired_size <= 0` means the queue is full?

Reply:
Yes. `desired_size` can even go to negative number, which means that -N messages were pushed while the queue was full